lib/bookmark_query.rb

```diff
@@ -27,54 +27,135 @@ def initialize(user:, guardian: nil, params: {})
+    # to polymorphic bookmakrks.
+    if SiteSetting.use_polymorphic_bookmarks
+      results = list_all_results_polymorphic(topics, pms)
+    else
+      results = list_all_results(topics, pms)
+    end
+
+    results = results.order(
+      "(CASE WHEN bookmarks.pinned THEN 0 ELSE 1 END),
+        bookmarks.reminder_at ASC,
+        bookmarks.updated_at DESC"
+    )
 
     if @params[:q].present?
-      term = @params[:q]
-      bookmark_ts_query = Search.ts_query(term: term)
-      results = results
-        .joins("LEFT JOIN post_search_data ON post_search_data.post_id = bookmarks.post_id")
-        .where(
-          "bookmarks.name ILIKE :q OR #{bookmark_ts_query} @@ post_search_data.search_data",
-          q: "%#{term}%"
-        )
+      if SiteSetting.use_polymorphic_bookmarks
+        results = polymorphic_search(results, @params[:q])
+      else
+        results = search(results, @params[:q])
+      end
     end
 
     if @page.positive?
       results = results.offset(@page * @params[:per_page])
     end
 
     results = results.limit(@limit)
+    BookmarkQuery.preload(results, self)
+    results
+  end
+
+  private
+
+  def list_all_results(topics, pms)
+    results = base_bookmarks.merge(topics.or(pms))
+    results = results.merge(Post.secured(@guardian))
+    results = @guardian.filter_allowed_categories(results)
 
     if BookmarkQuery.preloaded_custom_fields.any?
-      Topic.preload_custom_fields(
-        results.map(&:topic), BookmarkQuery.preloaded_custom_fields
-      )
+      Topic.preload_custom_fields(results, BookmarkQuery.preloaded_custom_fields)
     end
 
-    BookmarkQuery.preload(results, self)
-
-    @guardian.filter_allowed_categories(results)
+    results
   end
 
-  private
+  def list_all_results_polymorphic(topics, pms)
+    topic_results = topic_bookmarks.merge(topics.or(pms))
+
+    post_results = post_bookmarks.merge(topics.or(pms))
+    post_results = post_results.merge(Post.secured(@guardian))
+
+    topic_results = @guardian.filter_allowed_categories(topic_results)
+    post_results = @guardian.filter_allowed_categories(post_results)
+
+    if BookmarkQuery.preloaded_custom_fields.any?
+      Topic.preload_custom_fields(topic_results, BookmarkQuery.preloaded_custom_fields)
+      Topic.preload_custom_fields(post_results, BookmarkQuery.preloaded_custom_fields)
+    end
+
+    # TODO: At some point we may want to introduce ways for other Bookmarkable types
+    # to further filter results securely using merges, though this is not necessary just
+    # yet.
 
-  def user_bookmarks
-    # There is guaranteed to be a TopicUser record if the user has bookmarked
-    # a topic, see BookmarkManager
+    union_sql = "#{topic_results.to_sql} UNION #{post_results.to_sql}"
+    if Bookmark.registered_bookmarkables.any?
+      union_sql += " UNION #{other_bookmarks.to_sql}"
+    end
+    Bookmark.select("bookmarks.*").from("(#{union_sql}) as bookmarks").includes(bookmarkable: :bookmarkable_relation)
```

Comment:
I'm a little confused by the `.includes` at the end. It doesn't seem like anything is using the `bookmarkable_relation`. Also, we end up preloading `bookmarkable` twice when loading the bookmarks list because we do it once here and another time when we preload the polymorphic associations.

Reply:
Yeah good point I think I added this before the Preloader stuff, will remove.